### PR TITLE
Notes metadata migrated to coded terms field

### PR DESCRIPTION
**Pull request by bmquinn**
# Summary 

This PR modifies the `notes` work descriptive metadata field from a list of strings to a coded term field.

The new shape for the backend:
```elixir
%{
  notes: [
    %{note: "first note", type: %{id: "GENERAL_NOTE", scheme: "note_type"}},
    %{note: "second note", type: %{id: "BIOGRAPHICAL_HISTORICAL_NOTE", scheme: "note_type"}}
  ]
}
```
The new shape for the front end:
```js
[
  {
    note: "first note",
    type: {
      scheme: "NOTE_TYPE",
      id: "GENERAL_NOTE",
    },
  },
  {
    note: "second note",
    type: {
      scheme: "NOTE_TYPE",
      id: "BIOGRAPHICAL_HISTORICAL_NOTE",
    },
  },
];

```

# Specific Changes in this PR
- Adds `NoteEntry` embedded schema
- `WorkDescriptiveMetadata` schema updated to embed_many `NoteEntry`
- GraphQL API updated for work and batch queries/mutations
- Front end forms updated
- BatchEdit forms updated

# Version bump required by the PR

See [Semantic Versioning 2.0.0](https://semver.org/) for help discerning which is required.

- [ ] Patch
- [ ] Minor
- [x] Major

# Steps to Test

https://user-images.githubusercontent.com/1395707/144449854-cbf62c4a-7125-4302-bd79-13af01672f73.mp4

# Screenshots

Elasticsearch Work document:
elasticsearch

GraphQL work query:
workQuery

GraphQL update work mutation
updateWorkMutation

GraphQL code list query:
codeListQuery

# :rocket: Deployment Notes

**Note** - __if you check any of these boxes go to the [(always open) `main` <- `staging` PR](https://github.com/nulib/meadow/pulls) and add detailed notes and instructions to help out others who may end up deploying your changes to production__

- Backward compatible API changes
  - [ ] Database Schema changes
  - [ ] GraphQL API
  - [ ] Elasticsearch API
  - [ ] Ingest Sheet
  - [ ] CSV metadata export/update API
  - [ ] Shared Links export API
- Backwards-incompatible API changes
  - [x] Database Schema changes
  - [x] GraphQL API
  - [x] Elasticsearch API
  - [ ] Ingest Sheet
  - [x] CSV metadata export/update API
  - [ ] Shared Links export API
- [x] Requires data migration
- [ ] Requires database triggers disabled during deployment/migration
- [x] Requires reindex
- [ ] Terraform changes
  - [ ] Adds/requires new or changed Terraform variables
- [ ] Pipeline configuration changes (requires `mix meadow.pipeline.setup` run)
- [ ] Requires new variable added to `miscellany`
- [x] Specific deployment synchronization instructions with other apps/API's
  - [ ] DC needs to be updated to support the new shape of notes data
- [ ] Other specific instructions/tasks

# Tested/Verified
- [ ] End users/stakeholders

### CODE REVIEW

**Comment by mbklein** [comment]
I'm pretty sure all the changes here are backward _incompatible_ – i.e., clients written for the previous implementation will not be able to handle the database, GraphQL, or index in the new form.

File: assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx [hunk 1]
```diff
@@ -6,20 +6,30 @@ import {
 } from "../../../services/testing-helpers";
 import BatchEditAboutUncontrolledMetadata from "./UncontrolledMetadata";
 import { UNCONTROLLED_METADATA } from "../../../services/metadata";
+import { allCodeListMocks } from "@js/components/Work/controlledVocabulary.gql.mock";
+import { CodeListProvider } from "@js/context/code-list-context";
```
Comment (adamjarling) [inline assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx:10]: Thanks for using the relative path import syntax.

File: assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx [hunk 2]
```diff
@@ -6,20 +6,30 @@ import {
...
+      {
+        mocks: allCodeListMocks,
+      }
+    );
   });
 
-  it("renders the component", () => {
-    expect(screen.getByTestId("uncontrolled-metadata"));
+  it("renders the component", async () => {
+    expect(await screen.getByTestId("uncontrolled-metadata"));
```
Comment (adamjarling) [inline assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx:26]: `@testing-library/react` async selectors are kinda specific.  To get this to properly evaluate, you'll need to swap out `getByTestId` with `findByTestId`.  Anytime you front load the expression with `await`, you'll need to use the `findBy...`.

File: assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx [hunk 3]
```diff
@@ -6,20 +6,30 @@ import {
...
   });
 
-  it("renders expected uncontrolled metadata fields", () => {
+  it("renders expected uncontrolled metadata fields", async () => {
     for (let item of UNCONTROLLED_METADATA) {
-      expect(screen.getByTestId(item.name));
+      expect(await screen.getByTestId(item.name));
     }
+    expect(await screen.findByTestId("notes"));
```
Comment (adamjarling) [inline assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx:33]: This looks good.

File: assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx (hunk 2 above)
Comment (bmquinn) [reply assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx:26]: Didn't know that, thanks for the heads up! Will fix.

**Comment by kdid** [comment]
So, I had some existing data in there. I had a couple works with notes, most didn't. 
I pulled down your branch and ran `mix ecto.migrate`. 
Looking in the database the new format for the notes looked correct. But I'm seeing this error when index syncronization happens. (and when I manually ran hot swap) 
I'm not sure if I missed something?

```
[error] GenServer Meadow.Data.IndexWorker terminating
** (CaseClauseError) no case clause matching: {:error, [message: "is an invalid coded term for scheme NOTE_TYPE"]}
    (ecto 3.7.1) lib/ecto/schema/loader.ex:71: anonymous fn/5 in Ecto.Schema.Loader.unsafe_load/4
    (elixir 1.12.1) lib/enum.ex:2356: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir 1.12.1) lib/enum.ex:1553: Enum."-map/2-lists^map/1-0-"/2
    (ecto 3.7.1) lib/ecto/embedded.ex:70: Ecto.Embedded.load/3
    (ecto 3.7.1) lib/ecto/schema/loader.ex:71: anonymous fn/5 in Ecto.Schema.Loader.unsafe_load/4
    (elixir 1.12.1) lib/enum.ex:2356: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto 3.7.1) lib/ecto/embedded.ex:64: Ecto.Embedded.load/3
    (ecto 3.7.1) lib/ecto/type.ex:893: Ecto.Type.process_loaders/3
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:401: Ecto.Repo.Queryable.struct_load!/6
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:233: anonymous fn/5 in Ecto.Repo.Queryable.preprocessor/3
    (elixir 1.12.1) lib/enum.ex:1553: Enum."-map/2-lists^map/1-0-"/2
    (elixir 1.12.1) lib/enum.ex:1553: Enum."-map/2-lists^map/1-0-"/2
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:224: Ecto.Repo.Queryable.execute/4
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
    (meadow 4.4.3) lib/meadow/elasticsearch_diff_store.ex:31: Meadow.ElasticsearchDiffStore.retrieve/2
    (meadow 4.4.3) lib/meadow/data/indexer.ex:57: Meadow.Data.Indexer.synchronize_chunk/1
    (meadow 4.4.3) lib/meadow/data/indexer.ex:50: Meadow.Data.Indexer.synchronize_schema/1
    (elixir 1.12.1) lib/enum.ex:930: Enum."-each/2-lists^foreach/1-0-"/2
    (meadow 4.4.3) lib/meadow/data/indexer.ex:38: Meadow.Data.Indexer.synchronize_index/0
    (meadow 4.4.3) lib/meadow/data/index_worker.ex:13: Meadow.Data.IndexWorker.synchronize/1
Last message: :interval_task
State: %{interval: 120000, override: true, status: :running, timer_ref: #Reference<0.1449720804.38273032.170754>}
[info]  IntervalTask: Invoking Elixir.Meadow.Data.IndexWorker.synchronize every 120000ms
```

**Comment by adamjarling** [comment]
@bmquinn Big high five.  The front end code surrounding this code list stuff is no walk in the park, and you did an incredible job implementing in line with existing patterns in place.  I just hope the patterns are legit and hold up over time:). But the more we see of this type of data (notes, relatedUrl), perhaps we can create a new wrapper data type to handle in display and forms.

File: assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx (hunk 2 above)
Comment (bmquinn) [reply assets/js/components/BatchEdit/About/UncontrolledMetadata.test.jsx]: @adamjarling pushed the `findByTestId` version, thanks for the feedback 😄

**Comment by kdid** [comment]
Screen Shot 2021-12-03 at 7 56 45 AM

@bmquinn - I think the batch edit screens need the choice to add/replace like the other metadata has?
When I performed a performed a batch edit it just replaced the other notes I had. Is there some reason it's not present for. notes?

**Comment by davidschober** [comment]
I think that was a miss on related URL. It looks like @bmquinn fixed it for notes (at least I'm seeing it)
Screen Shot 2021-12-03 at 10 02 37 AM

I'll kick off an issue to add the dropdown to `related url`

**Comment by kdid** [comment]
@davidschober are you looking at staging? That's the old notes field. (for the string). This PR not merged.

**Comment by bmquinn** [comment]
@davidschober that notes field is the previous incarnation, but this PR actually removes that dropdown since the functionality is based off the `related_url` implementation. The two fields should be fixed together I'd wager.

**Comment by kdid** [comment]
@davidschober - I agree with @bmquinn - the two fields (NOTES and RELATED URL) should be fixed together in a separate piece of work/PR

**Review by kdid — APPROVED** [review]

**Pull request merged**